### ISSUE TITLE
Change working directory for `npm run scriptname` in package.json

### ISSUE THREAD

**Issue by eljefedelrodeodeljefe** [npm/npm#10957]
## TL; DR

> To increase `npm`'s capabilities as a build to, it could utilise `node`'s `child_process` cwd-option better. This would increase better platform support for building purposes also.

## Synopsis

This PR / feature request adds the above mentioned feature by checking whether the package
JSON includes a script in the form of...

```js
{
  "scripts": {
    "change-cwd-test": {
       "cwd": "test/fixtures/",
       "run": "node -e \"console.log(process.cwd())\""
    }
   }
}
```
## Motivation

This would allow for for enabling `npm` to be more of a build tool. For example some programs such as make don't allow non-cwd runs. And in order to not `cd` to it, this is supposed to greatly help.

Think of `make` which has the `-C [dir]` flag, though, but it occurs for example in the CPython (extensive) build script, that some subtask don't get this flag passed, or other unexpected behaviours. Node's own `child_process` has good support for manipulating it's `cwd` and could be the basis for the ecosystem's best build capabilities.

Node library authors currently tend to write `build.js` files manually doing this PR's behaviour and execute it at the preinstall script run. 

## Description 

This PR requires a change in [read-package-json](https://github.com/npm/read-package-json), which - you know - parses the `package.json`, but currently it does this in a pretty simple, yet safe manner.  

`read-package-json` then would attach a un-enumerable array to the `scripts` object, which the `npm` client can check for during execution and change the working directory.

From the `npm` client's POV, this seems a little odd, since it needs to transform and check a lot. But in order to pass hard checks here and w/ `npm` and not to break, this seemed to be necessary.

## Caveats

* discussable syntax. Looks ugly to be honest.
* decrease in safety of both libraries
* greater surface area for bugs

## Pros

* see **Motivation**.
* one step making node and / or npm a better build tool
* would solve this [issue on Stackoverflow](http://stackoverflow.com/questions/30286498/change-working-directory-for-npm-scripts) (**Disclaimer**: it's mine) 

## Meta

Credit: @eljefedelrodeodeljefe
Reviewed-By:
PR-URL: npm#10958

**Comment by CrabDude** [comment]
:+1:  Unless I'm mistaken, there's no non-ugly way to due this with npm "scripts".

**Comment by eljefedelrodeodeljefe** [comment]
CrabDude :+1: :)

**Comment by iarna** [comment]
Closing in favor of discussion in the PR. https://github.com/npm/npm/issues/10958

**Comment by bminer** [comment]
Just posting here in case someone else finds it.  This is now supported by NPM using `$INIT_CWD`.  See docs for [npm-run-script](https://docs.npmjs.com/cli/run-script)
